source/lac/matrix_scaling.cc

```diff
@@ -0,0 +1,1373 @@
+      // we want to avoid saving the whole column scaling on each MPI rank
+      if (matrix.n() == matrix.m())
+        locally_owned_cols = locally_owned_rows;
+      else
+        locally_owned_cols =
+          Utilities::MPI::create_evenly_distributed_partitioning(
+            matrix.get_mpi_communicator(), matrix.n());
+
+      converged = false;
+      ghost_columns.clear();
+      ghost_columns.set_size(matrix.n());
+
+      row_scaling.reinit(locally_owned_rows.n_elements());
+      column_scaling.reinit(locally_owned_cols.n_elements());
+
+      row_scaling    = 1.0;
+      column_scaling = 1.0;
+
+      // Identify ghost columns
+      for (const auto row : locally_owned_rows)
+        {
+          for (auto it = matrix.begin(row); it != matrix.end(row); ++it)
+            {
+              if (!locally_owned_cols.is_element(it->column()))
+                ghost_columns.add_index(it->column());
+            }
+        }
+      partitioner.reinit(locally_owned_cols,
+                         ghost_columns,
+                         matrix.get_mpi_communicator());
+
+      ghost_column_owners =
+        Utilities::MPI::compute_index_owner(locally_owned_cols,
+                                            ghost_columns,
+                                            matrix.get_mpi_communicator());
+
+      switch (control.algorithm)
+        {
+          case MatrixScaling::AdditionalData::ScalingAlgorithm::sinkhorn_knopp:
+            {
+              do_sk_scaling(matrix,
+                            control.sinkhorn_knopp_parameters.max_iterations);
+            }
+            {
+              break;
+            }
+
+          case MatrixScaling::AdditionalData::ScalingAlgorithm::
+            l1_linf_symmetry_preserving:
+            {
+              if (control.l1linf_parameters.start_inf_norm_steps > 0 &&
+                  !converged)
+                do_linfty_scaling(
+                  matrix, control.l1linf_parameters.start_inf_norm_steps);
+              if (control.l1linf_parameters.l1_norm_steps > 0 && !converged)
+                do_l1_scaling(matrix, control.l1linf_parameters.l1_norm_steps);
+              if (control.l1linf_parameters.end_inf_norm_steps > 0 &&
+                  !converged)
+                do_linfty_scaling(matrix,
+                                  control.l1linf_parameters.end_inf_norm_steps);
+            }
+            {
+              break;
+            }
+        }
+    }
+  else
+    {
+      (void)matrix;
+      Assert(false, ExcNotImplemented());
+    }
+
+  return converged;
+}
+
+
+
+template <class Matrix, class VectorType>
+bool
+MatrixScaling::scale_linear_system(Matrix &matrix, VectorType &rhs)
+{
+  AssertDimension(matrix.m(), rhs.size());
+  AssertDimension(matrix.m(), matrix.n());
+
+  if constexpr (is_supported_matrix<Matrix>() ||
+                std::is_same_v<Matrix, TrilinosWrappers::SparseMatrix> ||
+                std::is_same_v<Matrix, PETScWrappers::MPI::SparseMatrix>)
```

Comment:
This is duplicating a condition you have in the `scale_matrix()` function. I think it would be useful to create an abbreviation for it (like you already did for the sequential case) or perhaps to simply call `scale_matrix()` unconditionally and let the error be created there instead -- I think that's the best solution, rather than second-guessing here whether `scale_matrix()` has in fact been implemented for a specific matrix type.